### DISCUSSION TITLE
Allow expressions to return variables

### DISCUSSION THREAD

**Discussion by arthuro555** [4ian/GDevelop#2548] [Enhancement ideas (and their technical discussions)]
## The problem:
Let's say I have a "command" behavior for text input objects that splits words by space into an array. How would I let the user access the array of words? I could make a GetArgument(index) expression, but maybe I want to get the user get the amount of arguments, or modify the arguments... And end up re-implementing an array :p. Another possibility would be making an action to set a scene variable to the computed array. That is pretty verbose and not the most intuitive though :/

## The solution:
Why not just return the array directly? The user would use it this way:
`VariableString(Object.Command::GetArguments()[0])`, or `For each child of Object.Command::GetArguments() do {}`...
This is IMO the most intuitive way, less verbose and is less work for extension developers B)

**Comment by 4ian** [comment]
Short answer is: Yes we could rework the extensions so that expressions of type `variable` can be defined (instead of just number and string) and improve the `ExpressionParser2` to handle function calls inside a `variable` expression (we would need by the way to consolidate `scenevar`, `globalvar` and `objectvar` into a single `variable` parameter type. It would also open the door to have latter support for local/temporary variables in events, rather than relying always on scene variables. It would also open the door to have a single actions rather than 3 when you're dealing with something using a variable).

For your example though, you could (and should) expose easy to use expressions/actions for the most common use case, and as you're saying also a way to actually set or get the entire array from a variable (so you give a escape hatch for more complex manipulations, without complexifying your interface). That sounds like a sensible design I think :)
(but this is a bit a contrived example, this kind of manipulation might be better suited outside of a text input behavior I think?)

But having expressions returning variables would help for extensions/behaviors holding complex structures.
